metrics/inspector_prioritization.py

```diff
@@ -0,0 +1,274 @@
+
+
+def retrieve_road_segment_data(segments, gis):
+    segments_layer = FeatureLayer(SEGMENTS_FEATURE_SERVICE_URL, gis)
+    segment_ids = list(segments["PROPERTYRSN"].unique())
+    segment_batches = batch_list(segment_ids)
+    segment_data = []
+    for segment_batch in segment_batches:
+        segment_batch = ", ".join(map(str, segment_batch))
+        query_result = segments_layer.query(
+            where=f"segment_id in ({segment_batch})",
+            out_fields="*",
+            return_geometry=True,
+        )
+        segment_data.append(query_result.df)
+    segment_data = pd.concat(segment_data)
+    segments = segments.merge(
+        segment_data, left_on="PROPERTYRSN", right_on="SEGMENT_ID", how="inner"
+    )
+
+    # Flagging segments if they are in the DAPCZ for later scoring
+    dapcz_segments = retrieve_dapcz_segments(gis, segments_layer)
+    segments["is_dapcz"] = segments["SEGMENT_ID"].isin(dapcz_segments)
+
+    # Tagging (primary) segments with the appropriate ROW inspector zone
+    inspector_zones = FeatureLayer(ROW_INSPECTOR_SERVICE_URL, gis)
+    primary_segments = segments[segments["IS_PRIMARY"]]
+    primary_segments.drop_duplicates(subset=["SEGMENT_ID"], inplace=True)
+    primary_segments = primary_segments.to_dict(orient="records")
+    row_inspector_lookup = {}
+    for segment in primary_segments:
+        geom = segment["SHAPE"]
+        row_inspector_lookup[
+            segment["SEGMENT_ID"]
+        ] = retrieve_row_inspector_segments_by_segment(geom, gis, inspector_zones)
+    segments["row_inspector_zone"] = segments["SEGMENT_ID"].map(row_inspector_lookup)
+
+    return segments
+
+
+def retrieve_dapcz_segments(gis, segments_layer):
+    """
+    Gets the list of street segments within the Downtown Project Coordination Zone (DAPCZ) polygon.
+    """
+    dapcz_features = FeatureLayer(DAPCZ_FEATURE_SERVICE_URL, gis)
+    dapcz_features = dapcz_features.query(where="1=1", return_geometry=True)
+    dapcz_segments = []
+    for polygon in dapcz_features.features:
+        filter = intersects(polygon.geometry)
+        response = segments_layer.query(geometry_filter=filter, out_fields="SEGMENT_ID")
+        dapcz_segments.append(response.df)
+    dapcz_segments = pd.concat(dapcz_segments)
+    return list(dapcz_segments["SEGMENT_ID"])
+
+
+def retrieve_row_inspector_segments_by_segment(geom, gis, inspector_zones):
+    """
+    Tags roadway segments with appropriate ROW inspector zone.
+    If a segment is in multiple, only the last ROW inspector zone is used.
+    """
+    filter = intersects(geom)
+    response = inspector_zones.query(
+        geometry_filter=filter,
+        out_fields="ROW_INSPECTOR_ZONE_ID",
+        return_geometry=False,
+        gis=gis,
+    )
+    if response.features:
+        # It is possible a segment could intersect multiple zones, but we only take the first one.
+        return response.features[0].attributes["ROW_INSPECTOR_ZONE_ID"]
+    return None
```

Comment:
this all seems to be working fine! if this needed to run at a high frequency, or if the AGOL API were giving us trouble, it would be cool to write an ETL that makes a big ole lookup table of all the segments by inspector zone and dapcz intersection. this ETL could pull from the lookup table, since it probably would not need to be updated very often.